### PR TITLE
quic: use sds for upstream http/3

### PR DESCRIPTION
**Pull request by alyssawilk**
Risk Level: Medium (some data plane refactors, mostly no-ops for !HTTP/3)
Testing: turned up HTTP/3 upstream SDS integration tests
Docs Changes: n/a
Release Notes: n/a
part of #14829

### CODE REVIEW

File: source/common/quic/client_connection_factory_impl.cc [hunk 1]
```diff
@@ -19,19 +19,34 @@ getContext(Network::TransportSocketFactory& transport_socket_factory) {
   auto* quic_socket_factory =
       dynamic_cast<QuicClientTransportSocketFactory*>(&transport_socket_factory);
   ASSERT(quic_socket_factory != nullptr);
-  ASSERT(quic_socket_factory->sslCtx() != nullptr);
   return quic_socket_factory->sslCtx();
 }
 
+std::shared_ptr<quic::QuicCryptoClientConfig> PersistentQuicInfoImpl::cryptoConfig() {
```
Comment (danzh2010) [inline source/common/quic/client_connection_factory_impl.cc:25]: This is somewhat laid-back updating. If a client connection is created before the SSL context update, will its crypto config not be updated till another createQuicNetworkConnection() is called?

Comment (alyssawilk) [reply source/common/quic/client_connection_factory_impl.cc:25]: I don't think once a quic client is created we support reloading, so AFIK it'd only be applied when we created a new client anyway.

Comment (danzh2010) [reply source/common/quic/client_connection_factory_impl.cc:25]: ACK. 
Another question is if we need to copy over resumption tickets stored in EnvoyQuicSessionCache before replacing crypto config?

Comment (alyssawilk) [reply source/common/quic/client_connection_factory_impl.cc:25]: I would think if we changed client certs we'd not want to keep resumption info?  cc @RyanTheOptimist

Comment (RyanTheOptimist) [reply source/common/quic/client_connection_factory_impl.cc:25]: Yup, that sounds right to me.

File: source/common/http/http3/conn_pool.cc [hunk 2]
```diff
@@ -64,7 +65,13 @@ allocateConnPool(Event::Dispatcher& dispatcher, Random::RandomGenerator& random_
                  Upstream::ClusterConnectivityState& state, TimeSource& time_source) {
   return std::make_unique<Http3ConnPoolImpl>(
       host, priority, dispatcher, options, transport_socket_options, random_generator, state,
-      [](HttpConnPoolImplBase* pool) {
+      [](HttpConnPoolImplBase* pool) -> ::Envoy::ConnectionPool::ActiveClientPtr {
+        // If there's no ssl context, the secrets are not loaded. Fast-fail by returning null.
+        if (dynamic_cast<Quic::QuicClientTransportSocketFactory*>(
```
Comment (ggreenway) [inline source/common/http/http3/conn_pool.cc:70]: Is this cast guaranteed to succeed? Should it be a static_cast, and maybe an ASSERT that the dynamic_cast succeeds?

File: source/common/quic/quic_transport_socket_factory.h [hunk 3]
```diff
@@ -119,10 +117,8 @@ class QuicClientTransportSocketFactory : public QuicTransportSocketFactoryBase {
   }
 
 protected:
-  void onSecretUpdated() override {
-    // fallback_factory_ will update the stats.
-    // TODO(14829) Client transport socket factory may also need to update quic crypto.
-  }
+  // fallback factory will update the context.
```
Comment (ggreenway) [inline source/common/quic/quic_transport_socket_factory.h:120]: nit: capitalize Fallback or change back to `fallback_factory_`

**Review by ggreenway — CHANGES_REQUESTED** [review]
Looks good.

**Review by ggreenway — APPROVED** [review]

**Pull request merged**